keps/sig-api-machinery/3156-http3/README.md

```diff
@@ -0,0 +1,228 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+One of the main advantages of HTTP/3 is increased performance, specifically around fetching multiple
+objects simultaneously. With HTTP/2, any interruption (packet loss) in the TCP connection blocks all
+streams (Head of line blocking).
+
+In addition, HTTP/3 offers 0-RTT support, which means that subsequent connections can start up much
+faster by eliminating the TLS acknowledgement from the server when setting up the connection. This
+means the client can start requesting data much faster than with a full TLS negotiation.
+
+## Motivation
+
+Kubernetes clusters defaulted to HTTP/2 for the control plane communications to solve all the
+performance and reliability problems caused by the previous protocol HTTP/1.
+However, there are some notable exceptions:
+
+- [communication between apiserver and webhooks](https://github.com/kubernetes/kubernetes/pull/82090).
+- [communication that use SPDY or websockets](https://github.com/kubernetes/kubernetes/issues/7452)
+
+HTTP/3 solves some of the HTTP/2 problems, mainly the Head of Line Blocking problem because it uses
+QUIC RFC9000 for the transport layer, improving the performance and the resilience of the control
+plane.
+
+Despite the HTTP/3 protocol, as January 2022, is still an [Internet Draft](https://datatracker.ietf.org/doc/draft-ietf-quic-http/), it is already supported by 73% of running web browsers, and there is also a [golang implementation](https://github.com/lucas-clemente/quic-go) that is being used by known projects and companies.
+
+Having support for HTTP/3 will allow the project to have early feedback, and the possibility to influence it.
+
+### Goals
+
+- Support HTTP/3 on the apiserver wherever is possible
+- Offer an opt-in option in client-go to use HTTP/3
+
+### Non-Goals
+
+- Use or default any of the Kubernetes components to HTTP/3
+- Automatically upgrade connections to HTTP/3
+- Replace websockets or SPDY communications
+
+## Proposal
+
+The proposal is two fold:
+
+- support HTTP/3 in the apiserver, that will require to listen on the UDP port, in addition to current TCP port that is being used.
+
+- support HTTP/3 in client-go, so developers can consume it, that will require to add a new roundtripper and a new configuration option to opt-in.
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+- As a Kubernetes admin I'd like to be able to use HTTP/3 in front of my control-plane, so external users can benefit of its properties.
+
+#### Story 2
+
+- As a Kubernetes developer I'd like to be able to use HTTP/3 as my communication protocol in my operator, to be more resilient on environments with hostile network conditions.
```

Comment:
Maybe a story 3 about APIService or admission webhooks?

Reply:
what do you mean by APIService?

I don't want to include webhooks because adds an external dependency. Webhooks are already pinned to http/1 ... in summary I prefer to not open that can of worms :smile:

Reply:
> APIService

https://kubernetes.io/docs/tasks/extend-kubernetes/setup-extension-api-server/

I'm fine with either - we can list these things in non-goals.

Reply:
yeah, aggregated apiservers are other thing I'd like to avoid, I'll clarify it